### PR TITLE
Stabilse and update app

### PR DESCRIPTION
**Pull request by tom-rosewell**
Main Changes:

- Works with Capella
- Upgraded to use SDK 4.1.3
- Better code readability - comments and refactoring
- README parity with other sample apps
- Builds from a separate parent Docker image

Minor Changes:

- README aligns with style guidelines

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,25 +1,57 @@
...
-###### run local database, backend, and frontend
-#  $ docker-compose --profile local up         
-
-###### run local backend and frontend against a database in Couchbase Capella cloud
-#  $ CB_HOST=cb.{your-host}.cloud.couchbase.com  docker-compose --profile capella up
-
-###### run local database OR backend OR frontend in a mix-and match style
-#  $ docker-compose up db
-
+# -----------------------USAGE------------------------
```
Comment (osfameron) [inline docker-compose.yml:3]: nice improvement

File: travel.py [hunk 2]
```diff
@@ -1,28 +1,39 @@
...
+# USERNAME = 'Administrator'
+# PASSWORD = 'password'
+# ----------CAPELLA----------
+# CONNSTR = 'couchbases://db'
+# USERNAME = 'cbdemo'
+# PASSWORD = 'Password123!'
+# ---------------------------
+
+# Editing this file? Replicate your changes in the 'sample-app.py' file in
+# the 'docs-sdk-python' repo to have these changes appear in the tutorial.
```
Comment (osfameron) [inline travel.py:36]: This is a little hard to parse: may be worth prepending with "for Couchbase docs team"?
Also link to the actual file on github? I can't find it.

Also, why don't we tag the actual sample-app rather than copy/pasting? 🤔

File: travel.py [hunk 3]
```diff
@@ -243,45 +271,70 @@ def flightPaths(fromloc, toloc):
...
+                                     fromfaa=queryFrom, 
+                                     tofaa=queryTo, 
+                                     dayofweek=flightDay)
+
+        # The 'QueryResult' object can only be iterated over once - any further
+        # attempts to do so will result in an 'AlreadyQueried' exception. It is
+        # good practice to move the results into another data structure such as
+        # a list.
+        # Price data is not a part of the sample dataset, so a random number is
+        # picked and added to the result dict.
```
Comment (osfameron) [inline travel.py:327]: These comments are fab!

File: travel.py [hunk 4]
```diff
@@ -414,30 +469,32 @@ def signup(tenant):
...
 
+        userDocumentKey = lowercase(user)
+
+        agent = lowercase(tenant)
         scope = bucket.scope(agent)
         users = scope.collection('users')
 
-        querytype = "KV insert - scoped to {name}.users: document ".format(
-            name=scope.name)
+        queryType = f"KV insert - scoped to {scope.name}.users: document "
```
Comment (osfameron) [inline travel.py:483]: ooo, didn't know about f"strings"

File: Dockerfile [hunk 5]
```diff
@@ -1,20 +1,21 @@
...
 
 RUN apt-get update && apt-get install -y \
-    build-essential cmake \
-    git-all libssl-dev \
+    build-essential \
+    cmake \
+    libssl-dev \
     jq curl
 
-ADD . /app
```
Comment (osfameron) [inline Dockerfile:12]: Why is this removed?
Doesn't this mean that any *local* changes to the code won't be picked up if you're running with Docker?

File: README.md [hunk 6]
```diff
@@ -1,294 +1,210 @@
...
+You can still use Docker to run the Database and Frontend components if desired.
 
-```
-docker-compose up frontend
-```
+> **_NOTE:_** See above for specific details on running your database in Couchbase Capella.
 
-### Running the frontend manually
+1. Make sure you have `Python 3.7` or later installed on your machine.
+2. Copy the `requirements.txt` file from the sample-app-python folder in the [parent image repository](https://github.com/couchbaselabs/sample-app-parent-images) to the sample application directory
```
Comment (osfameron) [inline README.md]: This url is 404

Comment (osfameron) [reply README.md]: Copying the requirements.txt feels weird to me.
Worth discussing with Jared as local Python expert to see if it's an acceptable tradeoff for devs.

File: cached-backend.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,21 @@
+FROM python:3.9-slim-bullseye
```
Comment (osfameron) [inline cached-backend.Dockerfile:1]: in the compose file you have:

>   # cached-backend - Use the `sample-app-python` image
>  # from Docker Hub. The image has preinstalled all of
> # the dependencies, and contains a copy of the code.

This doesn't appear to use that image?

File: docker-compose.yml [hunk 8]
```diff
@@ -1,25 +1,57 @@
...
+#  $ docker compose up local-backend
+#
+# OR
+#
+#  $ docker compose up frontend
+#
+# Run developer tests against the backend:
+#  $ docker compose up test
+#
+# ----------------------PROFILES----------------------
```
Comment (osfameron) [inline docker-compose.yml:37]: This section is a good idea. It's a little hard to understand what the differences between some of the options are

server / local ?
capella / local-capella ?

Maybe a table? (awkward in ascii I know)
or some more explanation about the use-cases the different options support.

File: docker-compose.yml [hunk 9]
```diff
@@ -33,29 +65,80 @@ services:
...
+  # mounts the container on this repository.
+  #
+  # cached-capella-backend - Is the same image as the
+  # cached-backend, but uses the Capella user/pass.
+  #
+  # local-capella-backend - Functions the same as the
+  # local-backend, but uses the Capella user/pass.
+
+  cached-backend:
+    image: sample-app-python
```
Comment (osfameron) [inline docker-compose.yml]: ah, so you don't build the cached-backend.Dockerfile?

File: docker-compose.yml [hunk 10]
```diff
@@ -33,29 +65,80 @@ services:
...
     container_name: couchbase-sandbox-7.0.0
     profiles:
+      - server
       - local
       - mix-and-match
       - test
 
-  backend:
-    build: .
+  # ---------------------BACKENDS---------------------
```
Comment (osfameron) [inline docker-compose.yml:73]: This section may be necessary, but does kinda show how complicated this is 😄

File: docker-compose.yml [hunk 11]
```diff
@@ -66,9 +149,14 @@ services:
...
-    container_name: try-cb-api-cap
+    container_name: local-capella-backend-python
     profiles:
-      - capella
+      - local-capella
+
+  # ---------------------FRONTENDS---------------------
+  # There are 5 frontend services, all are functionally
+  # identical. The only difference is which profile the
+  # service belongs to, and which backend it depends on
```
Comment (osfameron) [inline docker-compose.yml:159]: Hmm. I think this complexity is the reason that I spun some things to mix-and-match.yml instead.

**Review by osfameron — COMMENTED** [review]
This is great work @tom-rosewell !

I'm a little uncertain about the combinatorial explosion of complexity in the Dockerfile.
Let's have a chat about this and see if we can tame it at all.
(we found with previous iterations that the Docker stuff needs so much more design thinking than you first imagine, and of course now the requirements are much more involved than previous)

It would also be worth speaking to a developer / DA / similar to get a feeling for it.

That said, let's have a quick chat, but this is great work with amazing attention to detail, and I'd be happy to approve and leave other stuff for a future iteration if need be.

File: README.md (hunk 6 above)
Comment (tom-rosewell) [reply README.md]: Forgot to update the README, most of that is out of date now. I'll get that done today.

File: docker-compose.yml (hunk 9 above)
Comment (tom-rosewell) [reply docker-compose.yml]: No, it's not used at all. The only reason it's there is if you wanted to update the image in Docker Hub (there's not really a better place to put it)

File: docker-compose.yml (hunk 10 above)
Comment (tom-rosewell) [reply docker-compose.yml:73]: This is probably only necessary for Python, the rest don't support Capella so there only need to be 2 backends
